tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -58,84 +62,86 @@ def jax_lanczos(matvec: Callable, arguments: List, init: jax.ShapedArray,
+      krylov_vectors = krylov_vectors.at[i, :].set(
+          jax.numpy.ravel(normalized_vector))
+      krylov_vectors = krylov_vectors.at[i + 1, :].set(
+          jax.numpy.ravel(next_vector))
+
+      return [krylov_vectors, alphas, betas, i + 1]
 
     def cond_fun(vals):
-      _, _, _, _, _, _, norm, threshold, iteration, maxiteration = vals
-
-      def check_thresh(check_vals):
-        val, thresh = check_vals
-        return jax.lax.cond(val < thresh, False, lambda x: x, True, lambda x: x)
-
-      return jax.lax.cond(iteration <= maxiteration, [norm, threshold],
-                          check_thresh, False, lambda x: x)
-    #TODO (mganahl): check if this runs on TPU (dtype issue)
-    numel = np.prod(init.shape).astype(np.int32)
-    krylov_vecs = jax.numpy.zeros((ncv + 1, numel), dtype=init.dtype)
-    norms = jax.numpy.zeros(ncv, dtype=init.dtype)
-    diag_elems = jax.numpy.zeros(ncv, dtype=init.dtype)
-
-    norms = jax.ops.index_update(norms, jax.ops.index[0], 1.0)
-
-    norms_dtype = jax.numpy.real(jax.numpy.empty((0, 0),
-                                                 dtype=init.dtype)).dtype
-    initvals = [
-        init, krylov_vecs, norms, diag_elems, matvec, arguments,
-        norms_dtype.type(1.0), landelta, 1, ncv
-    ]
-    output = jax.lax.while_loop(cond_fun, body_lanczos, initvals)
-    final_state, krylov_vecs, norms, diags, _, _, _, _, it, _ = output
-    krylov_vecs = jax.ops.index_update(krylov_vecs, jax.ops.index[it, :],
-                                       jax.numpy.ravel(final_state))
-
-    A_tridiag = jax.numpy.diag(diags) + jax.numpy.diag(
-        norms[1:], 1) + jax.numpy.diag(jax.numpy.conj(norms[1:]), -1)
+      betas, i = vals[-2], vals[-1]
+      norm = betas[i - 1]
+      return jax.lax.cond(i <= ncv, lambda x: x[0] > x[1], lambda x: False,
+                          [norm, landelta])
+
+    numel = np.prod(shape).astype(np.int32)
+    # note: ncv + 2 because the first vector is all zeros, and the
+    # last is the unnormalized residual.
+    krylov_vecs = jax.numpy.zeros((ncv + 2, numel), dtype=dtype)
+    # NOTE (mganahl): initial state is normalized inside the loop
+    krylov_vecs = krylov_vecs.at[1, :].set(jax.numpy.ravel(init))
+
+    # betas are the upper and lower diagonal elements
+    # of the projected linear operator
+    # the first two beta-values can be discarded
+    # set betas[0] to 1.0 for initialization of loop
+    # betas[2] is set to the norm of the initial state.
+    betas = jax.numpy.zeros(ncv + 1, dtype=dtype)
+    betas = betas.at[0].set(1.0)
+    # diagonal elements of the projected linear operator
+    alphas = jax.numpy.zeros(ncv, dtype=dtype)
+    initvals = [krylov_vecs, alphas, betas, 1]
+    krylov_vecs, alphas, betas, _ = jax.lax.while_loop(cond_fun, body_lanczos,
+                                                       initvals)
+    # FIXME (mganahl): if the while_loop stopps early at iteration i, alphas
```

Comment:
Another option would be to return the final iteration number to the user, who could then slice outside of Jitted code as needed.

Reply:
I added a warning. Generally yes, but I'd like to keep it for now. I finished implementing implicitly restarted lanczos in JAX, and this will become the new default anyways